### DISCUSSION TITLE
top() in the middle of a KQL query

### DISCUSSION THREAD

**Discussion by alightgoesout** [kdl-org/kdl#387] [Q&A]
The KQL grammar (on the v2 branch) allows the `top()` filter in the middle of a query. What should `a > top() > b` return ? I see several possibilities:

- return the same result as `top() > b` (after all, the spec says that top _returns all toplevel children of the current document_)
- return an empty set (the query does not make sense)
- return the same result as `a > b` (ignore `top()`, but that seems weird)
- fail the execution of the query (but I'd rather fail parsing instead)

**Comment by zkat** [comment]
If you look at the actual grammar on that branch, `top()` is only allowed at the very toplevel of a selector:

```abnf
query-str := $bom? query
query := selector q-ws* "||" q-ws* query | selector
selector := filter q-ws* selector-operator q-ws* selector | filter
selector-operator := ">>" | ">" | "++" | "+"
filter := "top(" q-ws* ")" | matchers
```
https://github.com/kdl-org/kdl/blob/kdl-v2/QUERY-SPEC.md#full-grammar

Unless I'm misreading my own grammar/spec wrong. Maybe what's missing is mentioning this fact further up in the spec?

**Reply by alightgoesout** [reply]
You're right I misread the grammar, my bad.

**Reply by larsgw** [reply]
I'm not sure that's the case? `selector` can expand to `filter q-ws* selector-operator q-ws* filter` right? I don't think this should be a syntax error though, it should return an empty set IMO.

**Reply by zkat** [reply]
oh yeah, you're right.

I definitely don't think we should have `top()` outside of the toplevel, though. I don't think we should just default to a random empty value.
